Possible false positive with `use-sequence-for-iteration` on in-place sets with dynamic content?

### Bug description

I'm getting a warning for `use-sequence-for-iteration` in my code, but I believe I have a valid reason to use it that should possibly be considered as an exception for the rule.

In my code I iterate over in-place sets with some in-place content, and some dynamic content to ensure that I iterate over the in-place content if it doesn't already exist in the dynamic content, while also ensuring I never iterate over the same thing twice.

Example:

```python
# test.py

def print_strings_and_hello(may_contain_hello):
    """
    Prints a set of strings without duplicates, while also ensuring 'hello' always gets
    printed
    """
    for string in {*may_contain_hello, "hello"}:
        print(string)


print_strings_and_hello(["hello", "goodbye"])  # prints both hello and goodbye
print_strings_and_hello(["goodbye"])  # prints both hello and goodbye

```

Do let me know if you think this is a bad practice!

I can't find much discussion about this warning online (which makes sense considering it was added recently) so I would love to engage and leave a paper trail for the next user who gets this warning for a similar usage.

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:8: [C0208(use-sequence-for-iteration), print_strings_and_hello] Use a sequence type when iterating over values

------------------------------------------------------------------
Your code has been rated at 8.00/10 (previous run: 8.00/10, +0.00)
```


### Expected behavior

No warning for the line

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.0
Python 3.9.2 (default, Apr 15 2021, 10:01:45) 
[Clang 12.0.0 (clang-1200.0.32.29)]
```

I think it is a false positive.

It looks like that case was explicitly contemplated:
https://github.com/PyCQA/pylint/blob/61b4db0dc31fd31fcfda36e4730cb44848af2d30/tests/functional/u/use/use_sequence_for_iteration.py#L16

But on the other hand, there was discussion that this warning was [only intended for in-place sets](https://github.com/PyCQA/pylint/issues/4776#issuecomment-889818180). This isn't really an "in place" set anymore if it's deduplicating an input.

I read the discussion on #4776, and I think this check was mostly about the nondeterministic ordering you get from iterating a set. Not about performance, really. But for performance, take a look at the most atomic case:
```
$ cat a.py
def deduplicate(list_in):
  for el in {*list_in}:
    print(el)

$ pylint a.py --disable=all --enable=use-sequence-for-iteration
************* Module a
a.py:2:12: C0208: Use a sequence type when iterating over values (use-sequence-for-iteration)

------------------------------------------------------------------
Your code has been rated at 6.67/10 (previous run: 0.00/10, +6.67)
```
User gets that and says, "okay pylint I trust you, let me cast to list after de-duping":
```python
>>> timeit('[x for x in var]', setup='var={1, 2, 3}')  # pretend var is dynamic
0.3215374090068508
>>> timeit('[x for x in list(var)]', setup='var={1, 2, 3}')
0.46824162200209685
```
Note that if the user does ⬆️ , then they will fall afoul of the checker proposed in #4355.

--- a/doc/whatsnew/fragments/5788.false_positive
+++ b/doc/whatsnew/fragments/5788.false_positive
@@ -1,0 +1,3 @@
+Fix ``use-sequence-for-iteration`` when unpacking a set with ``*``.
+
+Closes #5788

--- a/pylint/checkers/refactoring/recommendation_checker.py
+++ b/pylint/checkers/refactoring/recommendation_checker.py
@@ -338,7 +338,7 @@ class RecommendationChecker(checkers.BaseChecker):
         Sets using `*` are not considered in-place.
         """
         if isinstance(node.iter, nodes.Set) and not any(
-            isinstance(x, nodes.Starred) for x in node.iter.elts
+            utils.has_starred_node_recursive(node)
         ):
             self.add_message(
                 "use-sequence-for-iteration", node=node.iter, confidence=HIGH

--- a/pylint/checkers/refactoring/recommendation_checker.py
+++ b/pylint/checkers/refactoring/recommendation_checker.py
@@ -333,8 +333,13 @@ class RecommendationChecker(checkers.BaseChecker):
     def _check_use_sequence_for_iteration(
         self, node: nodes.For | nodes.Comprehension
     ) -> None:
-        """Check if code iterates over an in-place defined set."""
-        if isinstance(node.iter, nodes.Set):
+        """Check if code iterates over an in-place defined set.
+
+        Sets using `*` are not considered in-place.
+        """
+        if isinstance(node.iter, nodes.Set) and not any(
+            isinstance(x, nodes.Starred) for x in node.iter.elts
+        ):
             self.add_message("use-sequence-for-iteration", node=node.iter)
 
     @utils.only_required_for_messages("consider-using-f-string")

--- a/pylint/checkers/refactoring/recommendation_checker.py
+++ b/pylint/checkers/refactoring/recommendation_checker.py
@@ -9,7 +9,7 @@ from astroid import nodes
 
 from pylint import checkers
 from pylint.checkers import utils
-from pylint.interfaces import INFERENCE
+from pylint.interfaces import HIGH, INFERENCE
 
 
 class RecommendationChecker(checkers.BaseChecker):
@@ -340,7 +340,9 @@ class RecommendationChecker(checkers.BaseChecker):
         if isinstance(node.iter, nodes.Set) and not any(
             isinstance(x, nodes.Starred) for x in node.iter.elts
         ):
-            self.add_message("use-sequence-for-iteration", node=node.iter)
+            self.add_message(
+                "use-sequence-for-iteration", node=node.iter, confidence=HIGH
+            )
 
     @utils.only_required_for_messages("consider-using-f-string")
     def visit_const(self, node: nodes.Const) -> None:

--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -2029,6 +2029,18 @@ def find_assigned_names_recursive(
         for elt in target.elts:
             yield from find_assigned_names_recursive(elt)
 
+def has_starred_node_recursive(
+    node: nodes.For | nodes.Comprehension | nodes.Set,
+) -> Iterator[bool]:
+    """Yield ``True`` if a Starred node is found recursively."""
+    if isinstance(node, nodes.Starred):
+        yield True
+    elif isinstance(node, nodes.Set):
+        for elt in node.elts:
+            yield from has_starred_node_recursive(elt)
+    elif isinstance(node, (nodes.For, nodes.Comprehension)):
+        for elt in node.iter.elts:
+            yield from has_starred_node_recursive(elt)
 
 def is_hashable(node: nodes.NodeNG) -> bool:
     """Return whether any inferred value of `node` is hashable.

--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -2029,6 +2029,7 @@ def find_assigned_names_recursive(
         for elt in target.elts:
             yield from find_assigned_names_recursive(elt)
 
+
 def has_starred_node_recursive(
     node: nodes.For | nodes.Comprehension | nodes.Set,
 ) -> Iterator[bool]:
@@ -2041,6 +2042,7 @@ def has_starred_node_recursive(
     elif isinstance(node, (nodes.For, nodes.Comprehension)):
         for elt in node.iter.elts:
             yield from has_starred_node_recursive(elt)
+
 
 def is_hashable(node: nodes.NodeNG) -> bool:
     """Return whether any inferred value of `node` is hashable.

--- a/tests/functional/ext/code_style/cs_consider_using_tuple.py
+++ b/tests/functional/ext/code_style/cs_consider_using_tuple.py
@@ -28,4 +28,4 @@ for x in [2, *var]:  # [consider-using-tuple]
 
 # Don't emit warning for sets as this is handled by builtin checker
 (x for x in {1, 2, 3})  # [use-sequence-for-iteration]
-[x for x in {*var, 2}]  # [use-sequence-for-iteration]
+[x for x in {*var, 2}]

--- a/tests/functional/ext/code_style/cs_consider_using_tuple.txt
+++ b/tests/functional/ext/code_style/cs_consider_using_tuple.txt
@@ -4,5 +4,4 @@ consider-using-tuple:18:12:18:21::Consider using an in-place tuple instead of li
 consider-using-tuple:21:9:21:15::Consider using an in-place tuple instead of list:UNDEFINED
 consider-using-tuple:23:9:23:18::Consider using an in-place tuple instead of list:UNDEFINED
 consider-using-tuple:26:12:26:21::Consider using an in-place tuple instead of list:UNDEFINED
-use-sequence-for-iteration:30:12:30:21::Use a sequence type when iterating over values:UNDEFINED
-use-sequence-for-iteration:31:12:31:21::Use a sequence type when iterating over values:UNDEFINED
+use-sequence-for-iteration:30:12:30:21::Use a sequence type when iterating over values:HIGH

--- a/tests/functional/u/use/use_sequence_for_iteration.py
+++ b/tests/functional/u/use/use_sequence_for_iteration.py
@@ -13,4 +13,12 @@ for x in {1, 2, 3}:  # [use-sequence-for-iteration]
 [x for x in var]
 [x for x in {1, 2, 3}]  # [use-sequence-for-iteration]
 
-[x for x in {*var, 4}]  # [use-sequence-for-iteration]
+[x for x in {*var, 4}]
+
+def deduplicate(list_in):
+    for thing in {*list_in}:
+        print(thing)
+
+def deduplicate_two_lists(input1, input2):
+    for thing in {*input1, *input2}:
+        print(thing)

--- a/tests/functional/u/use/use_sequence_for_iteration.py
+++ b/tests/functional/u/use/use_sequence_for_iteration.py
@@ -22,3 +22,7 @@ def deduplicate(list_in):
 def deduplicate_two_lists(input1, input2):
     for thing in {*input1, *input2}:
         print(thing)
+
+def deduplicate_nested_sets(input1, input2, input3, input4):
+    for thing in {{*input1, *input2}, {*input3, *input4}}:
+        print(thing)

--- a/tests/functional/u/use/use_sequence_for_iteration.txt
+++ b/tests/functional/u/use/use_sequence_for_iteration.txt
@@ -1,3 +1,3 @@
-use-sequence-for-iteration:7:9:7:18::Use a sequence type when iterating over values:UNDEFINED
-use-sequence-for-iteration:11:12:11:21::Use a sequence type when iterating over values:UNDEFINED
-use-sequence-for-iteration:14:12:14:21::Use a sequence type when iterating over values:UNDEFINED
+use-sequence-for-iteration:7:9:7:18::Use a sequence type when iterating over values:HIGH
+use-sequence-for-iteration:11:12:11:21::Use a sequence type when iterating over values:HIGH
+use-sequence-for-iteration:14:12:14:21::Use a sequence type when iterating over values:HIGH

--- a/tests/functional/u/use/use_sequence_for_iteration.txt
+++ b/tests/functional/u/use/use_sequence_for_iteration.txt
@@ -1,4 +1,3 @@
 use-sequence-for-iteration:7:9:7:18::Use a sequence type when iterating over values:UNDEFINED
 use-sequence-for-iteration:11:12:11:21::Use a sequence type when iterating over values:UNDEFINED
 use-sequence-for-iteration:14:12:14:21::Use a sequence type when iterating over values:UNDEFINED
-use-sequence-for-iteration:16:12:16:21::Use a sequence type when iterating over values:UNDEFINED

